cmd/segment-reaper/observer.go

```diff
@@ -175,13 +175,37 @@ func (obsvr *observer) analyzeProject(ctx context.Context) error {
 			}
 
 			brokenObject := false
+			// expectedNumberOfSegments-1 because 'segments' doesn't contain last segment
 			if object.hasLastSegment {
 				if object.expectedNumberOfSegments == 0 {
 					if !object.segments.IsSequence() {
 						brokenObject = true
 					}
-				} else if object.segments.Count() != int(object.expectedNumberOfSegments)-1 {
-					// expectedNumberOfSegments-1 because 'segments' doesn't contain last segment
+				} else if object.segments.Count() < int(object.expectedNumberOfSegments)-1 {
+					brokenObject = true
+				} else if object.segments.Count() > int(object.expectedNumberOfSegments)-1 {
+					// verify if initial sequence is valid
+					for index := 0; index < int(object.expectedNumberOfSegments)-1; index++ {
+						has, err := object.segments.Has(index)
+						if err != nil {
+							panic(err)
+						}
+						if !has {
+							brokenObject = true
+						}
+					}
+					if !brokenObject {
+						// if initial sequence is valid then remove it from object and
+						// mark rest of segments as broken
+						for index := 0; index < int(object.expectedNumberOfSegments)-1; index++ {
```

Comment:
nit: we could have a method in bitmask for unsetting indexes up to some index or one that accepts a bottom and top range.

Reply:
agree but I would prefer to focus on testing and maybe later think about cleaning code

Reply:
Having that logic abstracted will make the testing of the analyze method easier, but we can do it later.

Reply:
I am in favor of simplifying the code in this method. This method is the most critical part of the zombie detection tool and we should do it right to avoid deleting valid data due to bugs.

Reply:
I'd would merge it, because there are many things going on top of others and do this change after this and the test is merged.
Obviously we could hold the release until we have done these changes.

I don't like to merge to master things like this, but considering that this is going to be an internal tool it should be fine.